<p align="center">
  <img src="docs/kids_first_logo.svg" alt="Kids First repository logo" width="660px" />
</p>
<p align="center">
  <a href="https://github.com/kids-first/kf-template-repo/blob/master/LICENSE"><img src="https://img.shields.io/github/license/kids-first/kf-template-repo.svg?style=for-the-badge"></a>
</p>

# Kids First single cell RNA Pipelines

This repo contains tools and workflows for analyzing single cell RNA (scRNA) data. It includes two workflows one for analyzing 10X data and one for analyzing Smart Seq 2 data.

The repo is currently in alpha phase.

## [10X workflow](docs/10X.md)

The workflow script that runs the tools is `workflows/kf_single_cell_10x.cwl`

The workflow runs [cellranger count](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/6.0/using/count),
on fastq files generated by the 10x single cell RNA workflow methodology.
Cell ranger count performs alignment, barcode counting, and filtering.
[SoupX](https://github.com/constantAmateur/SoupX) is used for subtraction of the RNA background.
[Scrublet](https://github.com/swolock/scrublet) is used to score and predict doublets.
Decontaminated outputs are aggregated using the [Seurat](https://satijalab.org/seurat/) R package from the Satija lab at the New York Genome Center.
A custom QC R markdown notebook developed by @AntoniaChroni is also run, which includes as it's main engine Seurat and [scooter](https://github.com/igordot/scooter)


## [Smart Seq 2 Workflow](docs/SMART_SEQ2.md)
The workflow uses [HISAT2](http://daehwankimlab.github.io/hisat2/) for alignment, [RNAseQC](https://github.com/getzlab/rnaseqc) to collect sequencing metrics, and [RSEM](https://deweylab.github.io/RSEM/) to calculate gene expression.
The outputs of the workflow are a matrix of cells gene counts in a loom file, a tarball containing a matrix of the gene counts in Matrix Market format, and a tsv with collected sequencing metrics from all cells analyzed.
Basic functionality includes removal of abnormal cells, dimensionality reduction, identification of differentially expressed features, and clustering.

Data are aligned to both the genome and the transcriptome. Genome aligned data are used to collect sequencing metrics and transcriptome aligned data are used to calculate expression.
