cwlVersion: v1.0
class: Workflow
id: kf_singe_cell_ss2
label: Kids First DRC single cell RNA Smart Seq 2 Workflow
doc: |-
  # KFDRC single cell RNA Smart Seq 2 workflow
  Workflow for aligning and counting single cell RNA reads generated by Smart Seq 2.

  ![data service logo](https://github.com/d3b-center/d3b-research-workflows/raw/master/doc/kfdrc-logo-sm.png)

requirements:
  - class: ScatterFeatureRequirement
  - class: MultipleInputFeatureRequirement
  - class: StepInputExpressionRequirement

inputs:
  final_output_basename: {type: string, doc: "Output basename for workflow output files"}
  #list of read1s
  fastq1s: {type: 'File[]', doc: "Array of fastq 1s to align"}
  #list of read2s
  fastq2s: {type: ['null', 'File[]?'], doc: "Array of fastq 2s to align"}
  #list of sample names
  sample_names: {type: 'string[]', doc: "Array of sample names"}
  #hisat genome reference
  hisat_genome_ref: {type: 'File', doc: "Hisat 2 genome reference"}
  #hisat transcriptome reference
  hisat_trans_ref: {type: 'File', doc: "Hisat 2 transcriptome reference"}
  #RNAseQC reference
  rnaseqc_gtf: {type: 'File', doc: "gtf file used by RNAseQC", sbg:suggestedValue: {class: 'File', path: '5d8bb21fe4b0950c4028f852', name: 'gencode.v27.primary_assembly.RNAseQC.gtf'}}
  #rsem reference
  rsem_reference: {type: 'File', doc: "RSEM reference file", sbg:suggestedValue: {class: 'File', path: '5d8bb21fe4b0950c4028f851', name: 'RSEM_GENCODE27.tar.gz'}}
  #ram and cpu
  cpus: { type: 'int?', default: 4, doc: "CPUs to allocate to call task"}
  ram: { type: 'int?', default: 8, doc: "RAM to allocate to call task in gb"}
  #option to determine strandedness... (might be able to steal this from bulk RNA seq)
  #option to output indiv bams???

outputs:
  #summary_out: {type: File, outputSource: cellranger/output_summary}
  matrix_loom: {type: 'File', outputSource: merge_looms/output_file}
  alignment_metrics_report: {type: 'File', outputSource: merge_rnaseqc_results/output_file}
  #seurat output
  #indiv bams (if option is made and used)

steps:
  #graph of plan: https://whimsical.com/smartseq2-hisat-workflow-tULYHw4T1x8znNJsZbd4a

#might need extra steps to figure out sample names and stranded options

#will need to append _genome and _trans to sample_names before aligning...
  #or not if we're not returning the bams???

  build_fastq2_array:
    run: ../tools/build_fastq2_array.cwl
    in:
      fastq1s: fastq1s
      fastq2s: fastq2s
    out: [fastq2s]

#these jobs are all scattered on the input fastqs

  #hisat 2 align to genome
    #scatter on fastq inputs
    #how to scatter on read1 but keep read2 and sample name in mind????
    #outputs genome aligned bam
  hisat2_align_genome:
    run: ../tools/hisat2_align.cwl
    scatter: [fastq1, fastq2, output_basename, input_id]
    scatterMethod: dotproduct
    in:
      reference: hisat_genome_ref
      fastq1: fastq1s
      fastq2: build_fastq2_array/fastq2s
      output_basename: sample_names
      input_id: sample_names
      strict:
        valueFrom: ${return Boolean(false)}
      ram: ram
      cpus: cpus
    out: [log_file, met_file, bam, bai]

  #hisat 2 align to transcriptome
    #same issues as above
    #outputs transcriptome aligned bam
  hisat2_align_trans:
    run: ../tools/hisat2_align.cwl
    scatter: [fastq1, fastq2, output_basename, input_id]
    scatterMethod: dotproduct
    in:
      reference: hisat_trans_ref
      fastq1: fastq1s
      fastq2: build_fastq2_array/fastq2s
      output_basename: sample_names
      input_id: sample_names
      strict:
        valueFrom: ${return Boolean(true)}
      ram: ram
      cpus: cpus
    out: [log_file, met_file, bam, bai]

  #RNAseQC
    #runs on the genome aligned bams
    #outputs a report
  rnaseqc:
    run: ../tools/rnaseqc.cwl
    scatter: input_bam
    in:
      input_bam: hisat2_align_genome/bam
      collapsed_gtf: rnaseqc_gtf
      #strand: figure out strand
      ram: ram
      cpus: cpus
    out: [metrics, gene_TPM, gene_count, exon_count]

  #RSEM
    #runs on the transcriptome aligned bams
    #doesn't depend on RNAseQC, so how to run them as the same time?
  rsem:
    run: ../tools/rsem_calc_express.cwl
    scatter: [input_bam, output_basename]
    scatterMethod: dotproduct
    in:
      input_bam: hisat2_align_trans/bam
      reference: rsem_reference
      output_basename: sample_names
    out: [gene_out, isoform_out, cnt_out, model_out, theta_out]

  #Create single loom
    #runs on RSEM output
  make_single_loom:
    run: ../tools/make_single_loom.cwl
    scatter: [rsem_matrix, sample_name]
    scatterMethod: dotproduct
    in:
      rsem_matrix: rsem/gene_out
      sample_name: sample_names
    out: [loom_file]

#these jobs are single runs that run on all of the output

  #Merge metric files
    #runs on all metrics files produced by RNAseQC
  merge_rnaseqc_results:
    run: ../tools/merge_rnaseqc_results.cwl
    in:
      output_basename: final_output_basename
      metrics_files: rnaseqc/metrics
    out: [output_file]

  #Merge loom files
    #runs on all loom files
  merge_looms:
    run: ../tools/merge_looms.cwl
    in:
      output_basename: final_output_basename
      loom_files: make_single_loom/loom_file
    out: [output_file]

  #Seurat clustering
    #runs on the merged loom file
    #will require even more testing

$namespaces:
  sbg: https://sevenbridges.com
hints:
  - class: 'sbg:maxNumberOfParallelInstances'
    value: 5
