# 10X workflow for estimating QC metrics

## Usage

To run `seurat_alignment_qc` Rscript in this module from the command line sequentially, use:

```
Rscript -e "rmarkdown::render('/WORK/tools/kf-dev-single-cell-rna/scripts/seurat_alignment_qc.Rmd', clean = TRUE,
     params=list(scooter_path='/scooter',
                 results_dir='TEST',
                 data_path='/WORK/volume/TEST_scRNA/pbmc_1k_v3',
                 sample_name='pbmc_1k_v3',
                 min_genes=200,
                 max_genes=6000,
                 max_mt=15,
                 normalize_method='log_norm'
                 ))"
                 
```

## Folder content

This folder contains scripts tasked to:
(1) Infer QC metrics and associated plots to visually explore the quality of the library
(2) Evaluate the QC metrics and set filters to remove low quality cells
in 10x single-cell- and single-nuclei-RNA-sequencing libraries (without cell hashing experiment).

The QC metrics will be divided in before and after filtering low quality cells from the library.


# A. Pre-Processing Information

## (1) QC metrics as generated by CellRanger count (count_summary: <sample_name>.web_summary.html)
## (2) Seurat QC metrics
[Seurat](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html) and [scooter](https://github.com/igordot/scooter) workflows are implemented to pre-process, filter and plot the RNA-sequencing data. CellRanger output is used for this step. User will have to define `params` as needed for their experiment: `scooter_path`, `results_dir`, `data_path`, `sample_name`, `min_genes`, `max_genes`, `max_mt`, `normalize_method`. 
  - Before and after filter: Plot distribution of the number of genes, UMI, and percent mitochondrial reads per cell.
  - Summary of Cell Statistics: Percent of reads in cells, Median UMI count per cell, Median genes detected per cell, Median percent reads mitochondrial.
  - Data were normalized by using the global-scaling normalization method “LogNormalize” that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. Then, the top 2000 highly variable genes (HVGs) are selected to subset features that indicate high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). Then, these HVGs are used as input to principal component analysis, and the top 30 principal components are selected. A combination of different dimensions (20, 25) and number of neighbors (30, 20, 10) are used along with the principal components to calculate the UMAP (Uniform Manifold Approximation and Projection) embeddings.

## (3) Estimating and filtering out ambient mRNA (`empty droplets`)

[SoupX method](https://cran.r-project.org/web/packages/SoupX/vignettes/pbmcTutorial.html) profiles “the soup”, i.e., collection of cell-free mRNAs floating in the input solution. The soup looks different for each input solution and strongly resembles the expression pattern obtained by summing all the individual cells.

SoupX calculates `Cell-specific contamination fraction` (Estimate (or manually set) the contamination fraction, the fraction of UMIs originating from the background, in each cell) and infers a `corrected expression matrix` (Correct the expression of each cell using the ambient mRNA expression profile and estimated contamination).

 CellRanger output is used for this step.
 -  Contamination summary table and Cell-specific contamination fraction plot are generated.

## (4) Estimating and filtering out doublets

Popular approach of scRNAseq uses oil droplets or wells to isolate single cells along with barcoded beads. Depending on the cell density loaded, a proportion of reaction volumes (i.e. droplets or wells) will capture more than one cell, forming ‘doublets’ (or ‘multiplets’), i.e. two or more cells captured by a single reaction volume and thus sequenced as a single-cell artifact. 

The proportion of doublets is proportional to the number of cells captured. Common in single-cell experiments to have 10-20% doublets, making accurate doublet detection critical.

Doublets are prevalent in single-cell sequencing data and can lead to artifactual findings. We will use a computational approach to calculate and remove doublets from the library.

Here, we use [Scrublet method](https://github.com/swolock/scrublet) for detecting doublets in RNA-sequencing data.


In either case, let us plot doublet prediction,, save table with predictions scores, and matrix after removing doublets.
 -  Summary table and histogram are generated.


## (5) Merging filtered data

Lastly, we merge count matrices from steps (2-4) after filtering out low quality cells, ambient RNA, and doublets. Seurat object and metadata for the library are saved to be used for downstream analyses.

