# 10X workflow for estimating QC metrics

## Usage

To run `seurat_alignment_qc` Rscript in this module from the command line sequentially, use:

```
Rscript -e "rmarkdown::render('/WORK/tools/kf-dev-single-cell-rna/scripts/seurat_alignment_qc.Rmd', clean = TRUE,
     params=list(scooter_path='/scooter',
                 results_dir='TEST',
                 data_path='/WORK/volume/TEST_scRNA/pbmc_1k_v3',
                 sample_name='pbmc_1k_v3',
                 min_genes=200,
                 max_genes=6000,
                 max_mt=15,
                 normalize_method='log_norm'
                 ))"
                 
```

## Folder content

This folder contains scripts tasked to:
(1) Infer QC metrics and associated plots to visually explore the quality of the library
(2) Evaluate the QC metrics and set filters to remove low quality cells
in 10x single-cell- and single-nuclei-RNA-sequencing libraries (without cell hashing experiment).

The QC metrics will be divided in before and after filtering low quality cells from the library.


# A. Pre-Processing Information

## (i) QC metrics as generated by CellRanger count (count_summary: <sample_name>.web_summary.html)
## (ii) Seurat QC metrics
  - User will have to define `params` as needed for their experiment: `scooter_path`, `results_dir`, `data_path`, `sample_name`, `min_genes`, `max_genes`, `max_mt`,      `normalize_method`
  - Before and after filter: Plot distribution of the number of genes, UMI, and percent mitochondrial reads per cell.
  - Summary of Cell Statistics: Percent of reads in cells, Median UMI count per cell, Median genes detected per cell, Median percent reads mitochondrial.


     


    AFTER FILTER
     - filtered_violin_qc: nFeature_RNA, nCount_RNA, percent.mt
     - filtered_qc: plot number of genes, plot number of UMIs, plot percent mitochondrial reads
     - Summary of Cell Statistics: Percent of reads in cells, Median UMI count per cell, Median genes detected per cell, Median percent reads mitochondrial
     - Cell Read Metrics Plots: Percent of cells expressing gene, Number of Expressed Genes Per Library Size (log10)
     - Normalize the data using `r normalize_method`, run `runpca`, run `dimred`/ Save metadata as a csv to `r glue("{out_path}/metadata_create_{sample_name}.csv")`/ Plot dimensionality reduction. / Save rds to `r paste0(out_path, "/", "seurat_obj.rds")`


- To add Knee plot: Total UMI count/Rank - Not sure how to do this: file:///Users/chronia/CHOP/projects/Alexslemonade/sample_data/SCPCS000022/SCPCL000022_qc.html



Then, we will use additional QC metrics to filter out ambient RNA and doublets in the library.

# B. Pre-Processing Information

## Estimating and filtering out ambient mRNA (`empty droplets`)

SoupX method profiles “the soup”, i.e., collection of cell-free mRNAs floating in the input solution. The soup looks different for each input solution and strongly resembles the expression pattern obtained by summing all the individual cells.

SoupX calculates `Cell-specific contamination fraction` (Estimate (or manually set) the contamination fraction, the fraction of UMIs originating from the background, in each cell) and infers a `corrected expression matrix` (Correct the expression of each cell using the ambient mRNA expression profile and estimated contamination).

This will read the `read_10x_data` from cellranger output (`./outs/analysis/clustering`) and should generate the below:

  - Contamination summary table
  - Plot Cell-specific contamination fraction 
  - Save filtered matrix


## Estimating and filtering out doublets

Popular approach of scRNAseq uses oil droplets or wells to isolate single cells along with barcoded beads. Depending on the cell density loaded, a proportion of reaction volumes (i.e. droplets or wells) will capture more than one cell, forming ‘doublets’ (or ‘multiplets’), i.e. two or more cells captured by a single reaction volume and thus sequenced as a single-cell artifact. 

The proportion of doublets is proportional to the number of cells captured. Common in single-cell experiments to have 10-20% doublets, making accurate doublet detection critical.

Doublets are prevalent in single-cell sequencing data and can lead to artifactual findings. We will use a computational approach to calculate and remove doublets from the library.

We can use scDblFinder (https://f1000research.com/articles/10-979) or Scrublet (https://github.com/swolock/scrublet).
In either case, let us plot doublet prediction,, save table with predictions scores, and matrix after removing doublets.



# C. Post-Processing Information
After filtering low quality cells, ambient RNA, and doublets, let us save the final seurat object and metadata for the library.
