---
title: "Seurat QC and Object for `r params$sample_name`"
author: 'Antonia Chroni <chronia@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
params:
  results_dir: './' # path to dir to save all output files
  data_path: '/path/to/10x/output/dir/filtered_feature_bc_matrix/'
  scooter_path: './' # File path to scooter directory
  sample_name: "pbmc_1k_v3"   # name of the sample in the library
  assay: "RNA"   # Assay 
  grouping: "orig.ident" # grouping for plots
  min_genes: 200 # minimum number of genes per cell
  mtDNA_pct_default: 5 # default pct mtDNA cutoff if miQC can't run
  normalize_method: "log_norm" # normalization method. One of log_norm or sct
  num_pcs: 30 # number of PCs to calculate
  num_dim: [20, 25] # number of PCs to use in UMAP
  num_neighbors: [30, 20, 10] # number of neighbors to use in UMAP
  prefix: "lognorm"   # prefix to add to UMAP. Useful if you are doing different normalizations, or using different subsets of the data
  log_file: "log" # log file
---
  
# Information about this notebook
This script creates Seurat object from the aligned count data and performs QC based on Seurat functions.
[miQC](https://bioconductor.org/packages/release/bioc/html/miQC.html) is used to filter out low quality cells in a given library.

For more information or updates, please see [Seurat](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html).

# Usage

To run the Rscript from the command line sequentially, use:

```
Rscript -e "rmarkdown::render('seurat_alignment_qc.Rmd', clean = TRUE,
      params=list(scooter_path='/scooter', 
                  results_dir='.', 
                  data_path='/path/to/cellranger/output', 
                  sample_name='pbmc_1k_v3',
                  min_genes=200,
                  normalize_method='log_norm',
                  num_pcs=30
                  ))"
```

# Set up
```{r load-library}
suppressPackageStartupMessages({
  library(future)
  library(cowplot)
  library(devtools)
  library(forcats)
  library(GGally)
  library(stringr)
  library(ggpmisc)
  library(ggrepel)
  library(miQC)
  library(scater)
  library(Seurat)
  library(SingleCellExperiment)
  library(flexmix) # to estimate mixtureModel for miQC
  library(irlba) # this solves the issue with RunUMAP code chunk

theme_set(theme_bw())

# evaluate Seurat R expressions asynchronously when possible (such as ScaleData) using future package
plan("multisession", workers = 4)
# increase the limit of the data to be shuttled between the processes from default 500MB to 50GB
options(future.globals.maxSize = 30 * 1024 ^ 3)
})
```

# Directories and paths to file Inputs/Outputs

```{r set-dir-and-file-names, echo=TRUE}
attach(params)
load_all(scooter_path)

# File path to data with the filtered_feature_bc_matrix
cat(paste(c("", sprintf("`%s`", list.files(paste(data_path, "/outs"), full.names = TRUE))),
          collapse = "\n- "))

cat("\n\n")

# File path to output data
results_dir <- 
  file.path(results_dir, paste0("Seurat_QC-", sample_name, "/"))
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

# File path to scratch directory
scratch_dir <- file.path(results_dir, "dr")
if (!dir.exists(scratch_dir)) {
  dir.create(scratch_dir)
}


knitr::opts_chunk$set(echo = FALSE,
                      warning = FALSE, 
                      message = FALSE, 
                      fig.path = results_dir,
                      fig.height = 10, 
                      fig.width = 15,
                      dev = c("png"))
```


# Read in 10x data

```{r read-in-data, echo=TRUE}
# Load counts to a list of matrices
counts_mat <- load_sample_counts_matrix(path = data_path,
                                        sample_name = sample_name)

# Create seurat object using gene expression data
seurat_obj <- create_seurat_obj(counts_matrix = counts_mat[["Gene Expression"]],
                                assay = "RNA",
                                log_file = NULL) 

# Save raw Seurat object
saveRDS(seurat_obj, file = paste0(results_dir, "/", "seurat_obj_raw.rds"))

# How many cells are in the seurat object?
cells_before_filter_num <- ncol(seurat_obj)
```

There are `r cells_before_filter_num` cells in the raw seurat object.

# Plot data before filter
Plot distribution of the number of genes, UMI, and percent mitochondrial reads per cell. 

```{r qc-before-filter, echo=TRUE}
# Plot number of genes 
genes_unfilt <- plot_distribution(seurat_obj, 
                                  features = "nFeature_RNA", 
                                  grouping = grouping) +
  geom_hline(yintercept = min_genes, color = "#D41159", size = 2) +
  #geom_hline(yintercept = max(seurat_obj@meta.data$nFeature_RNA), color = "#D41159", size = 2) +
  scale_fill_manual(values = alpha(c("#10559a"), .4)) +
  #annotate(geom="text", x=0.7, y = max(seurat_obj@meta.data$nFeature_RNA), label= glue(""), color="#D41159") +
  annotate(geom="text", x=0.7, y = min_genes - 200, label = glue("Min genes: {min_genes}"), color="#D41159")

# plot number of UMIs
umi_unfilt <- plot_distribution(seurat_obj, 
                                features = "nCount_RNA",
                                grouping = grouping) +
  scale_fill_manual(values = alpha(c("#10559a"), .4)) 


# plot percent mitochondrial reads
mito_unfilt <- plot_distribution(seurat_obj, 
                                 features = "pct_mito",
                                 grouping = grouping) +
  #geom_hline(yintercept = max(seurat_obj@meta.data$pct_mito), color = "#D41159", size = 2) +
  scale_fill_manual(values = alpha(c("#10559a"), .4)) 
  #annotate(geom="text", x=0.7, y = max(seurat_obj@meta.data$pct_mito), label = glue(""), color="#D41159")

# get the legend for one of the plots to use as legend for the combined plot
legend_grid <- get_legend(mito_unfilt)

title <- ggdraw() + 
  draw_label("Unfiltered-data", fontface = 'bold', x = 0, hjust = 0) +
  theme(
    # add margin on the left of the drawing canvas,
    # so title is aligned with left edge of first plot
    plot.margin = margin(0, 0, 0, 7))

# Combine plots 
plot_row <- plot_grid(genes_unfilt + theme(legend.position = "none"),
                      umi_unfilt + theme(legend.position = "none"),
                      mito_unfilt + theme(legend.position = "none"),
                      legend_grid,
                      ncol = 4)

plot_grid(title, plot_row, ncol = 1, rel_heights = c(0.1, 1))
```

# miQC steps for filtering low quality cells

We will use miQC to identify low-quality cells. miQC R package jointly models proportion of reads belonging to mitochondrial genes and number of unique genes detected. Cells with a high likelihood of being compromised (greater than 0.75) and cells that do not pass a minimum number of unique genes detected threshold of 200 will be removed from the counts matrix object.

## Create sce

```{r create-sce, echo=TRUE}
sce <- as.SingleCellExperiment(seurat_obj)
```

## Scater preprocessing
In order to calculate the percent of reads in a cell that map to mitochondrial genes, we first need to establish which genes are mitochondrial. For genes listed as HGNC symbols, this is as simple as searching for genes starting with mt-. For other IDs, we recommend using a biomaRt query to map to chromosomal location and identify all mitochondrial genes.

```{r scater-preprocessing, echo=TRUE}
# Set seed
set.seed(2023)

# Identify cells with mtDNA in the library
#mt_genes <- grepl("^mt-",  rownames(sce))
mt_genes <- grepl(sce$pct_mito, rownames(sce))
feature_ctrls <- list(mito = rownames(sce)[mt_genes])
# miQC is designed to be run with the Bioconductor package scater, which has a built-in function addPerCellQC to calculate basic QC metrics like number of unique genes detected per cell and total number of reads. When we pass in our list of mitochondrial genes, it will also calculate percent mitochondrial reads.

sce <- addPerCellQC(sce, subsets = feature_ctrls)
head(colData(sce))
```

## miQC

```{r miQC-plot, echo=TRUE}
print(plotMetrics(sce))

run_qc <- function(sce) {
  
  if (any(sce$subsets_mito_percent == 0)) {
    print("miQC did not work, use default mtDNA_pct_default for filtering.")
    filtered_sce <- subset(seurat_obj, subset = pct_mito < mtDNA_pct_default)
    
    } else {
      print("miQC worked.")
      model <- mixtureModel(sce)
      print(is.nan(model@logLik))                                    # Apply is.nan function
      parameters(model)
      head(posterior(model))
      
      # Plot
      # The cells at the very top of the graph are almost certainly compromised, 
      # most likely to have been derived from the distribution with fewer unique genes and higher baseline mitochondrial expression.
      print(plotModel(sce, model))

      # Plot
      print(plotFiltering(sce, model))

      # Convert sce back to seurat object 
      filtered_sce <- as.Seurat(filterCells(sce, model), counts = "counts",
                        data = "logcounts")

    }
    return (filtered_sce)
  }
  

# Run QC function
p <- run_qc(sce) 

```
## Filter data by min_genes

```{r data-filter, echo=TRUE}
# Now let's filter data by min_genes
seurat_obj <- subset(p, subset = nFeature_RNA > min_genes)

# How many cells are in the seurat object after filtering?
cells_so_after_filter_num <- ncol(seurat_obj)
```

There are `r cells_so_after_filter_num` cells in the seurat object after filtering.

## Number of cells express a specific gene

```{r freq-mean-after-filter, echo=TRUE}
# For each gene, calculate % of cells expressing it and the mean
pct_cells <- apply(as.data.frame(seurat_obj@assays$RNA@counts), 1, function(x) (sum(x != 0))) / ncol(seurat_obj@assays$RNA@counts)

gene_means <- rowMeans(as.data.frame(seurat_obj@assays$RNA@counts))

p <- as.data.frame(cbind(pct_cells, gene_means)) %>% 
  rownames_to_column("gene")

# Set seed
set.seed(2023)

ggplot(data = p, aes(x = gene_means, y = pct_cells, label = gene)) +
  geom_point(size = 3, alpha = 0.6) +
  xlab("Mean count per gene") +
  ylab("Percent cells expressing gene") +
  stat_dens2d_labels(geom = "text_repel", keep.fraction = 0.001)
```

## Number of expressed genes in the library

```{r cell-activity-after-filter, echo=TRUE}
# number of expressed genes
num_genes <- apply(as.data.frame(seurat_obj@assays$RNA@counts), 2, function(x) (sum(x != 0)))

libsize <- colSums(as.data.frame(seurat_obj@assays$RNA@counts))

l  <- as.data.frame(cbind(num_genes, libsize)) 

# Set seed
set.seed(2023)

ggplot(data = l, aes(x = log10(libsize), y = num_genes)) +
  geom_point(size = 3, alpha = 0.6) +
  xlab("Library Size (log10)") +
  ylab("Number of Expressed Genes") 
```

# Plot data after filter low quality cells

```{r qc-after-filter, echo=TRUE}
# plot number of genes
genes_sing <- plot_distribution(seurat_obj,
                                features = "nFeature_RNA",
                                grouping = grouping) +
  geom_hline(yintercept = min_genes, color = "#D41159", size = 2) +
  #geom_hline(yintercept = max(seurat_obj@meta.data$nFeature_RNA), color = "#D41159", size = 2) +
  scale_fill_manual(values = alpha(c("#10559a"), .4)) +
  #annotate(geom="text", x = 0.7, y = max(seurat_obj@meta.data$nFeature_RNA), label= glue(""), color="#D41159") +
  annotate(geom="text", x = 0.7, y = min_genes - 200, label=glue("Min genes: {min_genes}"), color="#D41159")


# plot number of UMIs
umi_sing <- plot_distribution(seurat_obj,
                              features = "nCount_RNA",
                              grouping = grouping) +
    scale_fill_manual(values = alpha(c("#10559a"), .4))


# plot percent mitochondrial reads
mito_sing <- plot_distribution(seurat_obj,
                               features = "pct_mito",
                               grouping = grouping) +
  #geom_hline(yintercept = max(seurat_obj@meta.data$pct_mito), color = "#D41159", size = 2) +
  scale_fill_manual(values = alpha(c("#10559a"), .4)) 
  #annotate(geom="text", x = 0.7, y = max(seurat_obj@meta.data$pct_mito), label = glue(""), color="#D41159")

# get the legend for one of the plots to use as legend for the combined plot
legend_grid <- get_legend(mito_sing)

# Combine plots
plot_grid(genes_sing + theme(legend.position = "none"),
          umi_sing + theme(legend.position = "none"),
          mito_sing + theme(legend.position = "none"),
          legend_grid,
          ncol = 4)
```

# Normalize data, find variable features, and scale the data

```{r normalize-variable-features-scale-data, echo=TRUE}
# normalize seurat object using specified method, on specified assay
seurat_obj <- NormalizeData(seurat_obj, normalization.method = "LogNormalize",
                             nfeatures = 2000, assay = "RNA")
seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", nfeatures = 2000) 
seurat_obj <- ScaleData(seurat_obj, features = VariableFeatures(seurat_obj))
```

# Run PCA

We will run PCA to reduce dimensionality of the matrix (as defined in the `params`).

```{r runpca, echo=TRUE}
# run PCA on specified assay using the number of PCs specified
seurat_obj <- run_dr(data = seurat_obj,
                     dr_method = "pca",
                     var_features = TRUE,
                     assay = assay,
                     num_pcs = num_pcs,
                     prefix =  prefix)
```
# Run UMAP

We will run UMAP based on the previous estimated PCs and for a variety of combinations as for the number of dimensions and neighbors (as defined in the `params`) to explore the clustering of each sample in the library.

```{r RunUMAP, echo=TRUE}
# Create a dataframe of all of the possible combinations of number of PCs 
# to use for UMAP, and the number of neighbors
num_dim_vect <- c(num_dim)
num_neighbors_vect <- c(num_neighbors)
possibilities <- expand.grid(num_dim_vect, num_neighbors_vect)

# For each of these combinations, calculate UMAP
for(i in 1:nrow(possibilities)) {
  num_dim <- possibilities[i, 1]
  num_neighbors <- possibilities[i, 2]
  seurat_obj <- run_dr(data = seurat_obj, dr_method = "umap", reduction = paste0("pca", prefix),
                     num_dim_use = num_dim, assay = "RNA", num_neighbors = num_neighbors,
                     prefix = glue("ndim{num_dim}nn{num_neighbors}{prefix}"))
}

# Generate metadata
reduction_names <- c(paste0("umap", "ndim", possibilities[,1], "nn", possibilities[,2], prefix), paste0("pca", prefix)) # Export the reductions to Seurat

metadata <- as_data_frame_seurat(seurat_obj, reduction = reduction_names,
                                 metadata = TRUE)
```

## Plot UMAP

```{r plot-UMAP, echo=TRUE}
# plot the first two PCs, and all of the different UMAPs
reduction_names <- c(paste0("UMAP", "ndim", possibilities[,1], "nn", possibilities[,2], prefix), paste0("PC", prefix))

plot_dr <- data.frame(X = paste0(reduction_names, "_1"),
                      Y = paste0(reduction_names, "_2"),
                      stringsAsFactors = FALSE)

for(i in 1:nrow(plot_dr)){
  print(current_plot <- plot_scatter(metadata = metadata,
                                     scratch_dir,
                                     proj_name = sample_name,
                                     log_file = log_file,
                                     X = plot_dr[i,1],
                                     Y = plot_dr[i,2],
                                     color = grouping,
                                     write = TRUE))
  
  print(current_plot <- plot_scatter(metadata = metadata,
                                     scratch_dir,
                                     proj_name = sample_name,
                                     log_file = log_file,
                                     X = plot_dr[i,1],
                                     Y = plot_dr[i,2],
                                     color = "nFeature_RNA",
                                     write = TRUE))
    
  print(current_plot <- plot_scatter(metadata = metadata,
                                     scratch_dir,
                                     proj_name = sample_name,
                                     log_file = log_file,
                                     X = plot_dr[i,1],
                                     Y = plot_dr[i,2],
                                     color = "nCount_RNA",
                                     write = TRUE))
    
  print(current_plot <- plot_scatter(metadata = metadata,
                                     scratch_dir,
                                     proj_name = sample_name,
                                     log_file = log_file,
                                     X = plot_dr[i,1],
                                     Y = plot_dr[i,2],
                                     color = "pct_mito",
                                     write = TRUE))
}

```

# Save filtered metadata and Seurat object

Let's save the filtered seurat object to be used for further downstream analysis (`seurat_obj.rds`).

```{r save-seurat-obj, echo=TRUE}
# Save metadata
write_tsv(metadata, path = glue("{results_dir}/metadata_create_{sample_name}.tsv"))

# Save Seurat object
saveRDS(seurat_obj, file = paste0(results_dir, "/", "seurat_obj.rds"))
```

```{r}
sessionInfo()
```

