---
title: "Final seurat object after filtering ambient RNA and doublets"
author: 'Antonia Chroni <chronia@chop.edu> for D3B'
date: "2023"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---

# Information about this notebook
This is a test for merging output results files from 3 qc steps: seurat_qc, soupx, and scrublet methods.

The data used here as ana example is the following CAVATICA task: https://cavatica.sbgenomics.com/u/d3b-bixu/kf-dev-single-cell-rna/tasks/951d22ae-c0f2-4449-8d2c-e583ba80816d/

See also this for cellranger output files: https://cavatica.sbgenomics.com/u/d3b-bixu/kf-dev-single-cell-rna/tasks/cf862908-847a-412b-8660-ae1157265e21/

# Set up
```{r load-library}
suppressPackageStartupMessages({
 # library(future) # this is for multiprocessing in seurat objects, if not needed let's remove this 
  library(Seurat)
  library(dplyr)
})
```

## Directories and paths to file Inputs/Outputs
```{r set-dir-and-file-names, echo=TRUE}
# Inputs
qc_data_path <- file.path("/Users/chronia/CHOP/GitHub/kf-dev-single-cell-rna/scripts/example/results/Seurat-QC-pbmc_1k_v3/seurat_obj.rds")
soupx_data_path <- file.path("/Users/chronia/CHOP/GitHub/kf-dev-single-cell-rna/scripts/example/data/prod_candidate.soupx.decontaminated_matrix.rds")
scrublet_path <- file.path("/Users/chronia/CHOP/GitHub/kf-dev-single-cell-rna/scripts/example/data/prod_candidate.scrublet.doublets.csv")

# File path to results directory
results_dir <-
  file.path("/Users/chronia/CHOP/GitHub/kf-dev-single-cell-rna/scripts/example", "results")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
```

# Load object from `seurat_qc` step

We used the Seurat package to run basic QC metrics and identify/remove doublets in the library. 

```{r load-seurat-qc-obj, echo=TRUE}
seurat_qc_obj <- readRDS(qc_data_path)

# How many cells are in the seurat qc object after removing contaminated cells?
cells_after_sqc_num <- ncol(seurat_qc_obj)
```

There are `r cells_after_sqc_num` cells in the object by the `seurat_qc` step.

# Remove ambient RNA

We used the SoupX package to detect ambient RNA in the library. We will create a new assay with the filtered data to be used in downstream analysis. This way we can preserve the initial matrix as generated by the `seurat_qc` step.

```{r Add-SoupX-count-mtx, echo=TRUE}
mtx <- readRDS(soupx_data_path)

soupx_cells <- colnames(seurat_qc_obj)[which(!(colnames(seurat_qc_obj) %in% colnames(mtx)))]
seurat_qc_obj <- subset(seurat_qc_obj, cells = soupx_cells, invert=TRUE)

mtx <- mtx[,colnames(seurat_qc_obj)]

seurat_qc_obj[["RNA_filter"]] <- CreateAssayObject(counts = mtx)
DefaultAssay(seurat_qc_obj) <- "RNA_filter"  # set RNA_filter assay so we use this matrix for any next qc steps

# How many cells are in the seurat object after removing contaminated cells?
cells_after_soupx_num <- ncol(seurat_qc_obj)
```

There are `r cells_after_soupx_num` cells in our current object after removing contaminated cells as identified in the `SoupX` step.

## Comments and future directions
In this specific example it appears that there were no cells removed after integrating the SoupX matrix. This indicates that there were zero cells identified as contamination. We need to figure out if this is correct. I checked the plot and it seems the rho_max is (close to) zero. 
Let's confirm that:
(1) we are using the correct matrix that doesn't contain any contaminated cells. 
(2) if we added the contamination summary table in the `SoupX` script, then let's save this file and see what's in there.


# Remove doublets
We used the Scrublet package to detect doublets in the library. 
Scrublet was run in the seurat aligned matrices before applying Seurat QC step.

```{r Scrublet, echo=TRUE}
scrublet <- read.csv(scrublet_path, sep = ",", header = FALSE, row.names=1)
colnames(scrublet) <- c("Doublet_score","Is_doublet") # set colnames

# Add scrublet results in the `seurat_qc_obj`
seurat_qc_obj <- AddMetaData(seurat_qc_obj, scrublet)

# How many doublets were identified by Scrublet?
scrublet_doublets <- seurat_qc_obj@meta.data %>% 
  filter(Is_doublet == "TRUE")

scrublet_doublets_num <- length(scrublet_doublets$Is_doublet)

# Let's filter the doublets out from the `seurat_qc_obj`
seurat_obj_filter <- subset(x = seurat_qc_obj, cells = WhichCells(seurat_qc_obj, expression = Is_doublet=="FALSE"))

# How many cells are in the seurat object after removing doublet cells identified by Scrublet?
cells_after_scrublet_num <- ncol(seurat_obj_filter)
```

There are `r scrublet_doublets_num` doublets in our current seurat_qc_obj as identified by Scrublet.
After removing these doublets, the final seurat_obj has `r cells_after_scrublet_num` cells. 

# Save object

The following final object can be used for downstream analyses.

```{r save-seurat-obj-filter, echo=TRUE}
saveRDS(seurat_obj_filter, file.path(results_dir, "seurat_obj_filter.rds"))
```

```{r}
sessionInfo()
```

