# 10X Alignment Workflow

The workflow script that runs the tools is `workflows/kf_STAR_Solo_10x_alignment.cwl`

The workflow runs [STAR Solo](https://github.com/alexdobin/STAR/blob/2.7.10b/docs/STARsolo.md),
on fastq files generated by the 10x single cell RNA workflow methodology.
Star Solo performs alignment, barcode counting, and filtering.
Quote from the STAR repo manual pdf:
>STARsolo output is designed to be a drop-in replacement for 10X CellRanger gene quantification output. It follows CellRanger logic for cell barcode whitelisting and UMI deduplication, and produces nearly identical gene counts in the same format. At the same time STARsolo is 10 times faster than the CellRanger.

Some custom tools are used to create Cell Ranger-like output file structure and h5 files for downstream refinement compatibility
A custom QC R markdown notebook developed by @AntoniaChroni is also run, which includes as it's main engine Seurat and [scooter](https://github.com/igordot/scooter)

## Software

- STAR Solo 2.7.10b
- Seurat 4.3.0.1
- SoupX 1.6.2

## Inputs
### multi-step
 - `output_basename`: basename used to name output files
 - `sample_name`: used as prefix for finding fastqs to analyze, e.g. 1k_PBMCs_TotalSeq_B_3p_LT_antibody if the names of the underlying fastqs are of the form 1k_PBMCs_TotalSeq_B_3p_LT_antibody_S1_L001_I1_001.fastq.gz, one per input fastq in the same order
### STAR Solo
 - `outSAMattrRGline`: Set if outputting bam, with TABS SEPARATING THE TAGS, format is: ID:sample_name LB:aliquot_id PL:platform SM:BSID for example ID:7316-242	LB:750189	PL:ILLUMINA	SM:BS_W72364MN
 - `genomeDir`: Tar gzipped reference that will be unzipped at run time
 - `readFilesIn1`: Input fastq file(s), gzipped or uncompressed
 - `readFilesIn2` R2 or 'mates' reads file(s), gzipped or uncompressed
 - `runThreadN` Num threads for STAR Solo to use, default: 16
 - `solo_type`: type of single-cell RNAseq
   - CB_UMI_Simple: (a.k.a. Droplet) one UMI and one Cell Barcode of fixed length in read2, e.g. Drop-seq and 10X Chromium
   - CB UMI_Complex: multiple Cell Barcodes of varying length, one UMI of fixed length and one adapter sequence of fixed length are allowed in read2 only (e.g. inDrop, ddSeq)
   - CB samTagOut: output Cell Barcode as CR and/or CB SAm tag. No UMI counting. –readFilesIn cDNA read1 [cDNA read2 if paired-end] CellBarcode read . Requires -outSAMtype BAM Unsorted [and/or SortedByCoordinate]
   - Smart-seq: each cell in a separate FASTQ (paired- or single-end), barcodes are corresponding read-groups, no UMI sequences, alignments deduplicated according to alignment start and end (after extending soft-clipped bases)
   - default: "CB_UMI_Simple"
 - `soloCBwhitelist`: file with whitelist of cell barcodes
 - `soloUMIlen`: UMI length, default: 12
 - `clipAdapterType`: adapter clipping type.
   - Hamming: adapter clipping based on Hamming distance, with the number of mismatches controlled by -clip5pAdapterMMp
   - CellRanger4: 5p and 3p adapter clipping similar to CellRanger4. Utilizes Opalpackage by Martin Sosic: https://github.com/Martinsos/opal
   - None: no adapter clipping, all other clip* parameters are disregarded
   - default: "CellRanger4"
 - `outFilterScoreMin`: alignment will be output only if its score is higher than or equal to this value, default: 30
 - `soloCBmatchWLtype`: matching the Cell Barcodes to the WhiteList.
   - Exact: only exact matches allowed
   - 1MM: only one match in whitelist with 1 mismatched base allowed. Allowed CBs have to have at least one read with exact match.
   - 1MM_multi: multiple matches in whitelist with 1 mismatched base allowed, posterior probability calculation is used choose one of the matches. Allowed CBs have to have at least one read with exact match. This option matches best with CellRanger 2.2.0
   - 1MM_multi_pseudocounts: same as 1MM Multi, but pseudocounts of 1 are added to all whitelist barcodes
   - 1MM_multi_Nbase_pseudocounts: same as 1MM multi pseudocounts, multimatching to WL is allowed for CBs with N-bases. This option matches best with CellRanger >= 3.0.0
   - EditDist_2: allow up to edit distance of 3 fpr each of the barcodes. May include one deletion + one insertion. Only works with -soloType CB UMI Complex. Matches to multiple passlist barcdoes are not allowed. Similar to ParseBio Split-seq pipeline
   - default: "1MM_multi_Nbase_pseudocounts
 - `soloUMIfiltering`: type of UMI filtering (for reads uniquely mapping to genes)
   - basic filtering: remove UMIs with N and homopolymers (similar to CellRanger 2.2.0)
   - MultiGeneUMI: basic + remove lower-count UMIs that map to more than one gene
   - MultiGeneUMI_All: basic + remove all UMIs that map to more than one gene
   - MultiGeneUMI_CR: basic + remove lower-count UMIs that map to more than one gene, matching CellRanger > 3.0.0. Only works with -soloUMIdedup 1MM_CR
   - default: "MultiGeneUMI_CR"
 - `soloUMIdedup`: type of UMI deduplication (collapsing) algorithm)
   - 1MM_All: all UMIs with 1 mismatch distance to each other are collapsed (i.e. counted once)
   - 1MM_Directional_UMItools: follows the ”directional” method from the UMI-tools by Smith, Heger and Sudbery (Genome Research 2017)
   - 1MM_Directional: same as 1MM Directional UMItools, but with more stringent criteria for duplicate UMIs
   - Exact: only exactly matching UMIs are collapsed
   - NoDedup: no deduplication of UMIs, count all reads
   - 1MM_CR: CellRanger2-4 algorithm for 1MM UMI collapsing
   - default: "1MM_CR"
 - `soloFeatures`: genomic features for which the UMI counts per Cell Barcode are collected
   - Gene: genes: reads match the gene transcript
   - SJ: splice junctions: reported in SJ.out.tab
   - GeneFull: full gene (pre-mRNA): full gene (pre-mRNA): count all reads overlapping genes' exons and introns
   - GeneFull_ExonOverIntron: count all reads overlapping genes' exons and introns: prioritize 100% overlap with exons
   - GeneFull_Ex50pAS: full gene (pre-RNA): count all reads overlapping genes' exons and introns: prioritize >50% overlap with exons. Do not count reads with 100% exonic overlap in the antisense direction
   - default: "GeneFull" 
 - `soloCellFilter`: cell filtering type and parameters
   - None: do not output filtered cells
   - TopCells: only report top cells by UMI count, followed by the exact number of cells
   - CellRanger2.2: simple filtering of CellRanger 2.2. Can be followed by numbers: number of expected cells, robust maximum percentile for UMI count, maximum to minimum ratio for UMI count. The harcoded values are from CellRanger: nExpectedCells=3000; maxPercentile=0.99; maxMinRatio=10
   - EmptyDrops_CR: EmptyDrops filtering in CellRanger flavor. Please cite the original EmptyDrops paper: A.T.L Lun et al, Genome Biology, 20, 63 (2019): https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1662-y, Can be followed by 10 numeric parameters: nExpectedCells maxPercentile maxMinRatio indMin indMax umiMin umiMinFracMedian candMaxN FDR simN. The harcoded values are from CellRanger: 3000 0.99 10 45000 90000 500 0.01 20000 0.01 10000"
   - default: "EmptyDrops_CR"
  outSAMtype: type of SAM/BAM output. None: no SAM/BAM output. Otherwise, first word is output type (BAM or SAM), second is sort type (Unsorted or SortedByCoordinate)
   - default: "None"
### seurat qc
 - `seurat_qc_min_genes`: minimum number of genes per cell
 - `seurat_qc_max_genes`: maximum number of genes per cell
 - `seurat_qc_max_mt`: maximum percent mitochondrial reads per cell
 - `seurat_qc_normalize_method`: normalization method. One of log_norm or sct
 - `seurat_qc_num_pcs`: number of PCs to calculate

### Outputs
 - `star_solo_counts_dir`: Tar gzipped counts output from STAR Solo
 - `star_solo_bam`: If flag given, aligned reads file
 - `star_solo_matrix_filtered`: Filtered counts matrix in Cell Ranger-style h5 format
 - `star_solo_matrix_raw`: Raw counts matrix in Cell Ranger-style h5 format
 - `star_solo_log_final_out`: STAR align summary stats
 - `star_solo_junctions`: STAR splice junction result file
 - `star_solo_cr_mimic_counts`: Tar ball of Cell Ranger-style counts dir from STAR Solo
 - `seurat_qc_html`: QC HTML Summary
 - `seurat_qc_rds`: QC rds. See docs for detailed contents of object
## QC RDS Output
Given that this is an rds file, for convenience we have posted the results of the R `str()` command on the output

```
Formal class 'Seurat' [package "SeuratObject"] with 13 slots
  ..@ assays      :List of 1
  .. ..$ RNA:Formal class 'Assay' [package "SeuratObject"] with 8 slots
  .. .. .. ..@ counts       :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  .. .. .. .. .. ..@ i       : int [1:5954964] 52 58 71 80 88 103 107 176 197 201 ...
  .. .. .. .. .. ..@ p       : int [1:3224] 0 1544 3756 5145 6744 9504 10322 11496 13834 17232 ...
  .. .. .. .. .. ..@ Dim     : int [1:2] 33425 3223
  .. .. .. .. .. ..@ Dimnames:List of 2
  .. .. .. .. .. .. ..$ : chr [1:33425] "DDX11L1" "WASH7P" "MIR1302-2HG" "ENSG00000238009" ...
  .. .. .. .. .. .. ..$ : chr [1:3223] "7316-2069:AAACCCACAGTCGTTA" "7316-2069:AAACCCAGTGCCTTTC" "7316-2069:AAACCCAGTGCGCTCA" "7316-2069:AAACGAAAGCAACAGC" ...
  .. .. .. .. .. ..@ x       : num [1:5954964] 1 3 1 5 1 4 5 13 21 4 ...
  .. .. .. .. .. ..@ factors : list()
  .. .. .. ..@ data         :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  .. .. .. .. .. ..@ i       : int [1:5954964] 52 58 71 80 88 103 107 176 197 201 ...
  .. .. .. .. .. ..@ p       : int [1:3224] 0 1544 3756 5145 6744 9504 10322 11496 13834 17232 ...
  .. .. .. .. .. ..@ Dim     : int [1:2] 33425 3223
  .. .. .. .. .. ..@ Dimnames:List of 2
  .. .. .. .. .. .. ..$ : chr [1:33425] "DDX11L1" "WASH7P" "MIR1302-2HG" "ENSG00000238009" ...
  .. .. .. .. .. .. ..$ : chr [1:3223] "7316-2069:AAACCCACAGTCGTTA" "7316-2069:AAACCCAGTGCCTTTC" "7316-2069:AAACCCAGTGCGCTCA" "7316-2069:AAACGAAAGCAACAGC" ...
  .. .. .. .. .. ..@ x       : num [1:5954964] 0.86 1.63 0.86 2.06 0.86 ...
  .. .. .. .. .. ..@ factors : list()
  .. .. .. ..@ scale.data   : num [1:2000, 1:3223] -0.0345 -0.0291 -0.0569 2.8124 -0.0868 ...
  .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. ..$ : chr [1:2000] "ENSG00000235146" "LINC02593" "TNFRSF4" "HES5" ...
  .. .. .. .. .. ..$ : chr [1:3223] "7316-2069:AAACCCACAGTCGTTA" "7316-2069:AAACCCAGTGCCTTTC" "7316-2069:AAACCCAGTGCGCTCA" "7316-2069:AAACGAAAGCAACAGC" ...
  .. .. .. ..@ key          : chr "rna_"
  .. .. .. ..@ assay.orig   : NULL
  .. .. .. ..@ var.features : chr [1:2000] "RMST" "TENM2" "OTX2-AS1" "HTR2C" ...
  .. .. .. ..@ meta.features:'data.frame':	33425 obs. of  5 variables:
  .. .. .. .. ..$ vst.mean                 : num [1:33425] 0.00031 0.02265 0.00155 0.00434 0.00279 ...
  .. .. .. .. ..$ vst.variance             : num [1:33425] 0.00031 0.02214 0.00155 0.00433 0.00279 ...
  .. .. .. .. ..$ vst.variance.expected    : num [1:33425] 0.00031 0.02855 0.00166 0.00484 0.00304 ...
  .. .. .. .. ..$ vst.variance.standardized: num [1:33425] 1 0.776 0.936 0.894 0.916 ...
  .. .. .. .. ..$ vst.variable             : logi [1:33425] FALSE FALSE FALSE FALSE FALSE FALSE ...
  .. .. .. ..@ misc         : list()
  ..@ meta.data   :'data.frame':	3223 obs. of  4 variables:
  .. ..$ orig.ident  : Factor w/ 1 level "7316-2069": 1 1 1 1 1 1 1 1 1 1 ...
  .. ..$ nCount_RNA  : num [1:3223] 7335 4379 2479 3162 6032 ...
  .. ..$ nFeature_RNA: int [1:3223] 1544 2212 1389 1599 2760 818 1174 2338 3398 2336 ...
  .. ..$ pct_mito    : num [1:3223] 0.191 0.206 0.081 0.063 0.066 0 0.146 0.144 0.105 0.04 ...
  ..@ active.assay: chr "RNA"
  ..@ active.ident: Factor w/ 1 level "7316-2069": 1 1 1 1 1 1 1 1 1 1 ...
  .. ..- attr(*, "names")= chr [1:3223] "7316-2069:AAACCCACAGTCGTTA" "7316-2069:AAACCCAGTGCCTTTC" "7316-2069:AAACCCAGTGCGCTCA" "7316-2069:AAACGAAAGCAACAGC" ...
  ..@ graphs      : list()
  ..@ neighbors   : list()
  ..@ reductions  :List of 7
  .. ..$ pcalognorm           :Formal class 'DimReduc' [package "SeuratObject"] with 9 slots
  .. .. .. ..@ cell.embeddings           : num [1:3223, 1:30] 4.46 5.31 3.56 3.27 4.71 ...
  .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. ..$ : chr [1:3223] "7316-2069:AAACCCACAGTCGTTA" "7316-2069:AAACCCAGTGCCTTTC" "7316-2069:AAACCCAGTGCGCTCA" "7316-2069:AAACGAAAGCAACAGC" ...
  .. .. .. .. .. ..$ : chr [1:30] "PClognorm_1" "PClognorm_2" "PClognorm_3" "PClognorm_4" ...
  .. .. .. ..@ feature.loadings          : num [1:2000, 1:30] 0.001694 0.002548 -0.000243 0.001501 0.000909 ...
  .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. ..$ : chr [1:2000] "RMST" "TENM2" "OTX2-AS1" "HTR2C" ...
  .. .. .. .. .. ..$ : chr [1:30] "PClognorm1" "PClognorm2" "PClognorm3" "PClognorm4" ...
  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ]
  .. .. .. ..@ assay.used                : chr "RNA"
  .. .. .. ..@ global                    : logi(0)
  .. .. .. ..@ stdev                     : num [1:30] 9.46 5.4 5.25 5.1 4.31 ...
  .. .. .. ..@ key                       : chr "PClognorm_"
  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package "SeuratObject"] with 4 slots
  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ]
  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ]
  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ]
  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ]
  .. .. .. ..@ misc                      : list()
  .. ..$ umapndim20nn30lognorm:Formal class 'DimReduc' [package "SeuratObject"] with 9 slots
  .. .. .. ..@ cell.embeddings           : num [1:3223, 1:2] 1.63 -0.62 -3.28 -2.71 -1.08 ...
  .. .. .. .. ..- attr(*, "scaled:center")= num [1:2] -1.142 -0.329
  .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. ..$ : chr [1:3223] "7316-2069:AAACCCACAGTCGTTA" "7316-2069:AAACCCAGTGCCTTTC" "7316-2069:AAACCCAGTGCGCTCA" "7316-2069:AAACGAAAGCAACAGC" ...
  .. .. .. .. .. ..$ : chr [1:2] "UMAPndim20nn30lognorm_1" "UMAPndim20nn30lognorm_2"
  .. .. .. ..@ feature.loadings          : num[0 , 0 ]
  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ]
  .. .. .. ..@ assay.used                : chr "RNA"
  .. .. .. ..@ global                    : logi(0)
  .. .. .. ..@ stdev                     : num(0)
  .. .. .. ..@ key                       : chr "UMAPndim20nn30lognorm_"
  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package "SeuratObject"] with 4 slots
  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ]
  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ]
  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ]
  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ]
  .. .. .. ..@ misc                      : list()
  .. ..$ umapndim25nn30lognorm:Formal class 'DimReduc' [package "SeuratObject"] with 9 slots
  .. .. .. ..@ cell.embeddings           : num [1:3223, 1:2] -3.46 -2.07 1.21 1.06 -1.26 ...
  .. .. .. .. ..- attr(*, "scaled:center")= num [1:2] 1.156 0.392
  .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. ..$ : chr [1:3223] "7316-2069:AAACCCACAGTCGTTA" "7316-2069:AAACCCAGTGCCTTTC" "7316-2069:AAACCCAGTGCGCTCA" "7316-2069:AAACGAAAGCAACAGC" ...
  .. .. .. .. .. ..$ : chr [1:2] "UMAPndim25nn30lognorm_1" "UMAPndim25nn30lognorm_2"
  .. .. .. ..@ feature.loadings          : num[0 , 0 ]
  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ]
  .. .. .. ..@ assay.used                : chr "RNA"
  .. .. .. ..@ global                    : logi(0)
  .. .. .. ..@ stdev                     : num(0)
  .. .. .. ..@ key                       : chr "UMAPndim25nn30lognorm_"
  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package "SeuratObject"] with 4 slots
  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ]
  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ]
  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ]
  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ]
  .. .. .. ..@ misc                      : list()
  .. ..$ umapndim20nn20lognorm:Formal class 'DimReduc' [package "SeuratObject"] with 9 slots
  .. .. .. ..@ cell.embeddings           : num [1:3223, 1:2] -1.646 0.505 2.244 -0.168 3.539 ...
  .. .. .. .. ..- attr(*, "scaled:center")= num [1:2] 0.774 -0.297
  .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. ..$ : chr [1:3223] "7316-2069:AAACCCACAGTCGTTA" "7316-2069:AAACCCAGTGCCTTTC" "7316-2069:AAACCCAGTGCGCTCA" "7316-2069:AAACGAAAGCAACAGC" ...
  .. .. .. .. .. ..$ : chr [1:2] "UMAPndim20nn20lognorm_1" "UMAPndim20nn20lognorm_2"
  .. .. .. ..@ feature.loadings          : num[0 , 0 ]
  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ]
  .. .. .. ..@ assay.used                : chr "RNA"
  .. .. .. ..@ global                    : logi(0)
  .. .. .. ..@ stdev                     : num(0)
  .. .. .. ..@ key                       : chr "UMAPndim20nn20lognorm_"
  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package "SeuratObject"] with 4 slots
  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ]
  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ]
  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ]
  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ]
  .. .. .. ..@ misc                      : list()
  .. ..$ umapndim25nn20lognorm:Formal class 'DimReduc' [package "SeuratObject"] with 9 slots
  .. .. .. ..@ cell.embeddings           : num [1:3223, 1:2] -2.552 -0.624 -0.685 0.48 2.847 ...
  .. .. .. .. ..- attr(*, "scaled:center")= num [1:2] -3.728 -0.223
  .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. ..$ : chr [1:3223] "7316-2069:AAACCCACAGTCGTTA" "7316-2069:AAACCCAGTGCCTTTC" "7316-2069:AAACCCAGTGCGCTCA" "7316-2069:AAACGAAAGCAACAGC" ...
  .. .. .. .. .. ..$ : chr [1:2] "UMAPndim25nn20lognorm_1" "UMAPndim25nn20lognorm_2"
  .. .. .. ..@ feature.loadings          : num[0 , 0 ]
  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ]
  .. .. .. ..@ assay.used                : chr "RNA"
  .. .. .. ..@ global                    : logi(0)
  .. .. .. ..@ stdev                     : num(0)
  .. .. .. ..@ key                       : chr "UMAPndim25nn20lognorm_"
  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package "SeuratObject"] with 4 slots
  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ]
  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ]
  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ]
  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ]
  .. .. .. ..@ misc                      : list()
  .. ..$ umapndim20nn10lognorm:Formal class 'DimReduc' [package "SeuratObject"] with 9 slots
  .. .. .. ..@ cell.embeddings           : num [1:3223, 1:2] -3.998 -4.553 -2.794 0.438 -8.244 ...
  .. .. .. .. ..- attr(*, "scaled:center")= num [1:2] -0.548 -0.39
  .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. ..$ : chr [1:3223] "7316-2069:AAACCCACAGTCGTTA" "7316-2069:AAACCCAGTGCCTTTC" "7316-2069:AAACCCAGTGCGCTCA" "7316-2069:AAACGAAAGCAACAGC" ...
  .. .. .. .. .. ..$ : chr [1:2] "UMAPndim20nn10lognorm_1" "UMAPndim20nn10lognorm_2"
  .. .. .. ..@ feature.loadings          : num[0 , 0 ]
  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ]
  .. .. .. ..@ assay.used                : chr "RNA"
  .. .. .. ..@ global                    : logi(0)
  .. .. .. ..@ stdev                     : num(0)
  .. .. .. ..@ key                       : chr "UMAPndim20nn10lognorm_"
  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package "SeuratObject"] with 4 slots
  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ]
  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ]
  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ]
  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ]
  .. .. .. ..@ misc                      : list()
  .. ..$ umapndim25nn10lognorm:Formal class 'DimReduc' [package "SeuratObject"] with 9 slots
  .. .. .. ..@ cell.embeddings           : num [1:3223, 1:2] 2.32 2.39 -1.04 -2.98 3.91 ...
  .. .. .. .. ..- attr(*, "scaled:center")= num [1:2] -0.51 0.719
  .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. ..$ : chr [1:3223] "7316-2069:AAACCCACAGTCGTTA" "7316-2069:AAACCCAGTGCCTTTC" "7316-2069:AAACCCAGTGCGCTCA" "7316-2069:AAACGAAAGCAACAGC" ...
  .. .. .. .. .. ..$ : chr [1:2] "UMAPndim25nn10lognorm_1" "UMAPndim25nn10lognorm_2"
  .. .. .. ..@ feature.loadings          : num[0 , 0 ]
  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ]
  .. .. .. ..@ assay.used                : chr "RNA"
  .. .. .. ..@ global                    : logi(0)
  .. .. .. ..@ stdev                     : num(0)
  .. .. .. ..@ key                       : chr "UMAPndim25nn10lognorm_"
  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package "SeuratObject"] with 4 slots
  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ]
  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ]
  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ]
  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ]
  .. .. .. ..@ misc                      : list()
  ..@ images      : list()
  ..@ project.name: chr "proj"
  ..@ misc        : list()
  ..@ version     :Classes 'package_version', 'numeric_version'  hidden list of 1
  .. ..$ : int [1:3] 4 1 3
  ..@ commands    :List of 3
  .. ..$ NormalizeData.RNA       :Formal class 'SeuratCommand' [package "SeuratObject"] with 5 slots
  .. .. .. ..@ name       : chr "NormalizeData.RNA"
  .. .. .. ..@ time.stamp : POSIXct[1:1], format: "2023-12-04 21:40:15"
  .. .. .. ..@ assay.used : chr "RNA"
  .. .. .. ..@ call.string: chr [1:2] "NormalizeData(seurat_obj, normalization.method = \"LogNormalize\", " "    nfeatures = 2000, assay = \"RNA\")"
  .. .. .. ..@ params     :List of 5
  .. .. .. .. ..$ assay               : chr "RNA"
  .. .. .. .. ..$ normalization.method: chr "LogNormalize"
  .. .. .. .. ..$ scale.factor        : num 10000
  .. .. .. .. ..$ margin              : num 2
  .. .. .. .. ..$ verbose             : logi TRUE
  .. ..$ FindVariableFeatures.RNA:Formal class 'SeuratCommand' [package "SeuratObject"] with 5 slots
  .. .. .. ..@ name       : chr "FindVariableFeatures.RNA"
  .. .. .. ..@ time.stamp : POSIXct[1:1], format: "2023-12-04 21:40:18"
  .. .. .. ..@ assay.used : chr "RNA"
  .. .. .. ..@ call.string: chr [1:2] "FindVariableFeatures(seurat_obj, selection.method = \"vst\", " "    nfeatures = 2000)"
  .. .. .. ..@ params     :List of 12
  .. .. .. .. ..$ assay              : chr "RNA"
  .. .. .. .. ..$ selection.method   : chr "vst"
  .. .. .. .. ..$ loess.span         : num 0.3
  .. .. .. .. ..$ clip.max           : chr "auto"
  .. .. .. .. ..$ mean.function      :function (mat, display_progress)
  .. .. .. .. ..$ dispersion.function:function (mat, display_progress)
  .. .. .. .. ..$ num.bin            : num 20
  .. .. .. .. ..$ binning.method     : chr "equal_width"
  .. .. .. .. ..$ nfeatures          : num 2000
  .. .. .. .. ..$ mean.cutoff        : num [1:2] 0.1 8
  .. .. .. .. ..$ dispersion.cutoff  : num [1:2] 1 Inf
  .. .. .. .. ..$ verbose            : logi TRUE
  .. ..$ ScaleData.RNA           :Formal class 'SeuratCommand' [package "SeuratObject"] with 5 slots
  .. .. .. ..@ name       : chr "ScaleData.RNA"
  .. .. .. ..@ time.stamp : POSIXct[1:1], format: "2023-12-04 21:40:21"
  .. .. .. ..@ assay.used : chr "RNA"
  .. .. .. ..@ call.string: chr "ScaleData(seurat_obj, features = VariableFeatures(seurat_obj))"
  .. .. .. ..@ params     :List of 10
  .. .. .. .. ..$ features          : chr [1:2000] "RMST" "TENM2" "OTX2-AS1" "HTR2C" ...
  .. .. .. .. ..$ assay             : chr "RNA"
  .. .. .. .. ..$ model.use         : chr "linear"
  .. .. .. .. ..$ use.umi           : logi FALSE
  .. .. .. .. ..$ do.scale          : logi TRUE
  .. .. .. .. ..$ do.center         : logi TRUE
  .. .. .. .. ..$ scale.max         : num 10
  .. .. .. .. ..$ block.size        : num 1000
  .. .. .. .. ..$ min.cells.to.block: num 3000
  .. .. .. .. ..$ verbose           : logi TRUE
  ..@ tools       : list()
```